homeassistant/components/roborock/vacuum.py

```diff
@@ -0,0 +1,233 @@
+from abc import ABC
+from typing import Any
+
+from roborock.code_mappings import FAN_SPEED_CODES
+from roborock.typing import RoborockCommand, RoborockDeviceInfo
+
+from homeassistant.components.vacuum import (
+    ATTR_BATTERY_ICON,
+    ATTR_FAN_SPEED,
+    ATTR_FAN_SPEED_LIST,
+    STATE_CLEANING,
+    STATE_DOCKED,
+    STATE_ERROR,
+    STATE_IDLE,
+    STATE_PAUSED,
+    STATE_RETURNING,
+    StateVacuumEntity,
+    VacuumEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_BATTERY_LEVEL, ATTR_STATE
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import slugify
+
+from . import RoborockDataUpdateCoordinator
+from .const import DOMAIN
+from .device import RoborockCoordinatedEntity
+
+STATE_CODE_TO_STATE = {
+    1: STATE_IDLE,  # "Starting"
+    2: STATE_IDLE,  # "Charger disconnected"
+    3: STATE_IDLE,  # "Idle"
+    4: STATE_CLEANING,  # "Remote control active"
+    5: STATE_CLEANING,  # "Cleaning"
+    6: STATE_RETURNING,  # "Returning home"
+    7: STATE_CLEANING,  # "Manual mode"
+    8: STATE_DOCKED,  # "Charging"
+    9: STATE_ERROR,  # "Charging problem"
+    10: STATE_PAUSED,  # "Paused"
+    11: STATE_CLEANING,  # "Spot cleaning"
+    12: STATE_ERROR,  # "Error"
+    13: STATE_IDLE,  # "Shutting down"
+    14: STATE_DOCKED,  # "Updating"
+    15: STATE_RETURNING,  # "Docking"
+    16: STATE_CLEANING,  # "Going to target"
+    17: STATE_CLEANING,  # "Zoned cleaning"
+    18: STATE_CLEANING,  # "Segment cleaning"
+    22: STATE_DOCKED,  # "Emptying the bin" on s7+
+    23: STATE_DOCKED,  # "Washing the mop" on s7maxV
+    26: STATE_RETURNING,  # "Going to wash the mop" on s7maxV
+    100: STATE_DOCKED,  # "Charging complete"
+    101: STATE_ERROR,  # "Device offline"
+}
+
+
+ATTR_STATUS = "status"
+ATTR_ERROR = "error"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Roborock sensor."""
+    coordinator: RoborockDataUpdateCoordinator = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+    entities = []
+    for device_id, device_info in coordinator.api.device_map.items():
+        unique_id = slugify(device_id)
+        entities.append(RoborockVacuum(unique_id, device_info, coordinator))
+    async_add_entities(entities)
+
+
+class RoborockVacuum(RoborockCoordinatedEntity, StateVacuumEntity, ABC):
+    """General Representation of a Roborock vacuum."""
+
+    def __init__(
+        self,
+        unique_id: str,
+        device: RoborockDeviceInfo,
+        coordinator: RoborockDataUpdateCoordinator,
+    ) -> None:
+        """Initialize a vacuum."""
+        StateVacuumEntity.__init__(self)
+        RoborockCoordinatedEntity.__init__(self, device, coordinator, unique_id)
+        self.manual_seqnum = 0
+        self._device = device
```

Comment:
The parent class takes every field of `_device` and copies it as a member variable. How about picking one apparoch other the other.  This appears to be unused.